### PR TITLE
[Refactoring] Strong types for ImapCommand

### PR DESCRIPTION
**Pull request by chibenwa**
- Strong typing of ImapCommand
 - Avoid mistakes in the request <-> command relationship

### CODE REVIEW

File: protocols/imap/src/main/java/org/apache/james/imap/decode/base/AbstractImapCommandParser.java [hunk 1]
```diff
@@ -67,15 +67,15 @@ protected StatusResponse taggedNo(Tag tag, ImapCommand command, HumanReadableTex
...
-            return statusResponseFactory.taggedBad(tag, command, e.getKey());
+        if (command.validForState(session.getState())) {
+            try {
+                return decode(command, request, tag, session);
+            } catch (DecodingException e) {
+                LOGGER.debug("Cannot parse protocol ", e);
+                return statusResponseFactory.taggedBad(tag, command, e.getKey());
+            }
         }
+        return statusResponseFactory.taggedNo(tag, command, HumanReadableText.INVALID_COMMAND);
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/base/AbstractImapCommandParser.java]: FYI, I personally prefer the previous version:
```
if (!validCondition) {
  return failFast;
}
normalBehaviour;
```

File: protocols/imap/src/main/java/org/apache/james/imap/api/ImapCommand.java [hunk 2]
```diff
@@ -61,21 +68,7 @@ public String getName() {
...
-            result = validInNonAuthenticated;
-            break;
-        case SELECTED:
-            result = validInSelected;
-            break;
-        default:
-            result = false;
-        }
-        return result;
+        return validity.validStates.contains(state);
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/ImapCommand.java]: I think we can have a domain method name for this, like: `validity.allowed(state)`

**Comment by Arsnael** [comment]
Needs a rebase !

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**